### PR TITLE
feat: cleanup PVS and autosave branches

### LINKED ISSUE #576: Remove PVS for older commits
As a follow-up to #573 , we may want to get rid of stale PVS.
We can imagine that a user won't need the PVS anymore when he starts a pod from a more recent commit. That means we could add to the `POST /servers` the logic to check if there are other PVS for the same repo that were created from an ancestor commit, or from a branch merged to the target branch.

### PR DESCRIPTION
**Pull request by olevski**
closes #576 

Note on terminology: I use PVS and pvc interchangeably below.

The logic is the same as discussed in #576, namely old autosave branches or pvcs are removed when:
- the commit where the new session is launched has any parent commit that has an autosave branch or pvc tied to it
- in the case of the pvc we check if this autosave pvc is part of a running pod and if not the pvc is deleted
- branches are similarly deleted - without the need for checking for running pods

In addition to the above the following changes were made:
- Autosave branches are created and used alongside with pvcs because this way we can let the user recover if they commit changes, but do not push them and then they go and shut down their session. In this case only the autosave branches can capture the changes because pvcs are deleted when user sessions are gracefully (i.e. purposefully) deleted. Also it is hard to check from the server side if a user session has a clean commit history or not when a pod is being deleted.
- Since even when pvcs are enabled autosave branches are still used (see bullet point above), the autosave deletion logic will now remove autosave branches if it finds a matching pvc.
- pre-stop script that creates the autosave branch was edited to use `git stash`. This is because in the current code this script does not revert to the original (dirty) state of the repo after it creates the autosave branch. This made perfect sense when there were no pvcs but now it does not. Now it does not make sense because the pvc is preferentially used to restore from an autosave rather than an autosave branch. And the creation of the autosave branch which is done in the pvc itself should not change the state of the pvc to something that is different than where the user left off.

Deployed at https://tasko.dev.renku.ch/ with support for pvcs.

/deploy

### CODE REVIEW

**Comment by lorenzo-cavazzi** [comment]
If we are going to use autosave branches all the time when the user gracefully stops from the UI (which is great in my opinion) and we delete the previous content when starting a new environment, then I think we should revisit how we save them.

Screenshot_20210420_145025

Something confusing for the user is how we associate the autosave with the starting commit. Let's say the user starts from commit `abcdef1` and creates more content (let's say commit `abcdef2`) and pushes. He keeps working on that and he accidentally stops the environment, creating an autosave with the non-pushed content `abcdef3`. Currently, the autosaved branch refers to `abcdef1` even if the user probably expects that to be attached to `abcdef2`.

The following is an example where I started the environment from `b1ab2e7d`, pushed 2 commits, and kept working locally adding other content. On the right, the autosave is bound to `b1ab2e7d`. The star identifying the autosave would be even harder to noticed if I pushed more commits in between (say 10 or 20).

Screenshot_20210420_150504

Even though that problem is independent of this PR, I think we should address it before merging. Otherwise, a user who doesn't notice the autosave associated with an old commit, would lose everything as soon as he starts a new environment from the latest commit.

Would it make sense to bind the commit to the latest pushed combination of branch/commit (with a preference for the current branch name in case of conflict) instead of using the starting branch/commit?

**Comment by olevski** [comment]
@lorenzo-cavazzi this is a very good find and it is easily fixable for the autosave branches. However once we change this for the autosave branches there is a difference between the naming of the autosave pvcs and branches. The autosave pvcs are always annotated with the initial commit that spawned the user session but the autosave branches naming follows whatever commits the user pushed as you suggested. It would be nice if we could do the same for the pvcs, however this is not straightforward and it would require continually patching the user pvc to keep track with all the commits that are pushed to the remote by the user. This is a problem because it is hard to get information about the local state of a user session on the server side where one would have to patch the pvc.

Because of this I think we should stick to what we currently have and explain this well in the docs and communicate it clearly in the UI. I am open to alternatives however - if you have any ideas on how to resolve this issue with the pvcs.

**Comment by olevski** [comment]
One thing that comes to mind is to have a separate dropdown/menu that shows all the commits that result in autosaves in the UI rather than have them both combined and distinguish with a `*`. If we have a separate list with autosaves we can remove the corresponding entry in the regular list so that we avoid duplicating things and we prevent the user from launching a 'fresh' session as long as an autosave exists. 

Another thing we can do is prevent the user from launching sessions from commits which are children of the autosave as long as the autosave branch or pvc is around. This way we force the user to resolve their unsaved work before continuing to launch new sessions.

**Comment by lorenzo-cavazzi** [comment]
I agree that having different behaviors between autosave branches and pvs may be confusing. If updating the branch/commit label attached to the pvs is not feasible at release time, then it's probably better to stay with the current solution for autosave branches, even though that will keep being confusing for users.

I'm not sure having a separate list and forcing the user to pick a commit from there leads to a better UX. On the other hand, I like the idea of showing a warning when starting from a more recent commit than the one with autosaved content. We still can't show much more details for pvs since the content is not accessible until the environment is up, but it's probably the best we can do.

**Comment by olevski** [comment]
So I wanted to summarize all the scenarios that this should cover and that I tested. Also I wanted to explain how the removal of old autosave pvcs or branches is done.

image

image

I added extra logic to the code that deletes autosaves from parent commits so that it checks how "far" these commits are from the commit where the new session is launched. If the commits are farther (i.e. older) than 5 commits then the autosave pvc and/or branch is deleted. Deleting an autosave branch or pvc tied to an immediate parent seemed a bit extreme.

Let me know if you agree with these changes.

**Comment by lorenzo-cavazzi** [comment]
I'm not sure the number of commits should play any role. Some people commit very often, others prefer to commit rarely -- though renku creates commits for you when using renku commands.
I would keep the logic simple and apply the same behavior, independent of how many commits there are in between the launched session and the reference for the saved session.

**Comment by olevski** [comment]
Ok I will remove the logic that looks at this 'commit distance' for removing autosave branches and PVs.

**Comment by olevski** [comment]
Ok I removed the commit distance requirement and tested in my dev delployment. Everything works.

**Review by ableuler — CHANGES_REQUESTED** [review]
Looks very good, works as advertised and carefully handles all possible scenarios I can think of. Just two suggestions for small improvements.

File: renku_notebooks/api/classes/server.py [hunk 1]
```diff
@@ -386,6 +387,75 @@ def get_logs(self, max_log_lines=0, container_name="notebook"):
...
+        def _get_all_mounted_pvcs():
+            pvcs = []
+            for pod in self._user.pods:
+                for volume in pod.spec.volumes:
+                    pvc = volume.persistent_volume_claim
+                    if pvc is not None:
+                        pvcs.append(pvc.claim_name)
+            return pvcs
+
+        def _has_parent(commit, parent, gl_project):
```
[the remark this answers is not in the document]
Comment (olevski) [reply renku_notebooks/api/classes/server.py:400]: Cool I did not know if this api endpoint. I will add the change.

File: renku_notebooks/api/classes/server.py [hunk 2]
```diff
@@ -246,13 +247,10 @@ def _get_start_payload(self):
...
-            if not pvc_exists:
-                self._create_pvc(
-                    storage_size=self.server_options.get("disk_request"),
-                    storage_class=current_app.config[
-                        "NOTEBOOKS_SESSION_PVS_STORAGE_CLASS"
-                    ],
-                )
+            self._create_pvc(
+                storage_size=self.server_options.get("disk_request"),
+                storage_class=current_app.config["NOTEBOOKS_SESSION_PVS_STORAGE_CLASS"],
```
[the remark this answers is not in the document]
Comment (olevski) [reply renku_notebooks/api/classes/server.py:252]: This is easy I will make the change

**Review by ableuler — APPROVED** [review]

**Pull request merged**